content/algorithms/dag/index.md

```diff
@@ -0,0 +1,362 @@
+
+Until the current level(`zero_indegree`) is empty, we generate the next one.
+
+We process all the vertices of the current level in variable `this_generation`
+and we store the next level in variable `zero_degree`.
+
+For each vertex from the current level(`this_generation`), we remove all of its outgoing edges.
+
+If the input degree of some vertex is zeroed, then we add it to the next level
+and remove it from the `indegree_map` dictionary.
+
+After we have processed the current level(`this_generation`), we can yield it.
+
+```
+while zero_indegree:
+    this_generation = zero_indegree
+    zero_indegree = []
+    for node in this_generation:
+        for child in G.neighbors(node):
+            indegree_map[child] -= 1
+
+            if indegree_map[child] == 0:
+                zero_indegree.append(child)
+                del indegree_map[child]
+
+    yield this_generation
+```
+
+#### Step 4. Graph contains a cycle.
+
+If, after the operation of the main cycle, there are still vertices in the graph, then there is a cycle in it.
+
+```
+if indegree_map:
+    raise nx.NetworkXUnfeasible(
+        "Graph contains a cycle or graph changed during iteration"
+    )
+```
+
+#### Addition 1. Topological sort is defined only on directed graphs.
+
+```
+if not G.is_directed():
+    raise nx.NetworkXError("Topological sort not defined on undirected graphs.")
+```
+
+#### Addition 2. Topological sort works on multigraphs as well.
+
+```
+multigraph = G.is_multigraph()
+```
+
+Replace
+
+```
+indegree_map[child] -= 1
+```
+
+with
+
+```
+indegree_map[child] -= len(G[node][child]) if multigraph else 1
+```
+
+#### Addition 3. Graph changed during iteration.
+
+Between passing different levels in a topological sort, the graph could change.
+
+#### Combine all steps.
+
+```{code-cell} ipython3
+def topological_generations(G):
+    if not G.is_directed():
+        raise nx.NetworkXError("Topological sort not defined on undirected graphs.")
+
+    multigraph = G.is_multigraph()
+    indegree_map = {v: d for v, d in G.in_degree() if d > 0}
+    zero_indegree = [v for v, d in G.in_degree() if d == 0]
+
+    while zero_indegree:
+        this_generation = zero_indegree
+        zero_indegree = []
+        for node in this_generation:
+            if node not in G:
+                raise RuntimeError("Graph changed during iteration")
+            for child in G.neighbors(node):
+                try:
+                    indegree_map[child] -= len(G[node][child]) if multigraph else 1
+                except KeyError as e:
+                    raise RuntimeError("Graph changed during iteration") from e
+                if indegree_map[child] == 0:
+                    zero_indegree.append(child)
+                    del indegree_map[child]
+        yield this_generation
+
+    if indegree_map:
+        raise nx.NetworkXUnfeasible(
+            "Graph contains a cycle or graph changed during iteration"
+        )
+```
```

Comment:
Rather than manually copy over the current implementation, it would be good if we could link to it instead in case it ends up changing. Two ideas for how to do this are:
 - Use GitHub, e.g. https://github.com/networkx/networkx/blob/main/networkx/algorithms/dag.py#L105-L179 (note that the `blob/main` part of the URL will guarantee that this link always points to the development version)
 - Try to find a way to render the source code directly in the notebook. E.g. in the normal notebook interface you could do `nx.topological_generations??`, but I don't think that will work in the html (worth a try though).